README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

A Flutter package provide a widget that show piano keyboard layout.

## Getting started

We support some default piano keyboard layout as well as custom layout

![](https://raw.githubusercontent.com/de-men/piano_widget/main/readme_images/example.gif)

## Usage

Please take a look at the (example)[https://github.com/de-men/piano_widget/example]

The widget support factories as some default (piano keyboard layout)[http://www.piano-keyboard-guide.com/piano-keyboard-layout.html]

```dart
ListView(
    padding: const EdgeInsets.all(16),
    children: [
        const Text('32 Keys start with C'),
        PianoWidget.keys32c(),
        const Divider(),
        const Text('32 Keys start with F'),
        PianoWidget.keys32f(),
        const Divider(),
        const Text('36 Keys start with C'),
        PianoWidget.keys36c(),
        const Divider(),
        const Text('36 Keys start with F'),
        PianoWidget.keys36f(),
        const Divider(),
        const Text('37 Keys start with C'),
        PianoWidget.keys37c(),
        const Divider(),
        const Text('37 Keys start with F'),
        PianoWidget.keys37f(),
        const Divider(),
        const Text('49 Keys'),
        PianoWidget.keys49(),
        const Divider(),
        const Text('54 Keys'),
        PianoWidget.keys54(),
        const Divider(),
        const Text('61 Keys'),
        PianoWidget.keys61(),
        const Divider(),
        const Text('76 Keys'),
        PianoWidget.keys76(),
        const Divider(),
        const Text('88 Keys'),
        PianoWidget.keys88(),
    ],
)
```

## Additional information

Feel free to give feedbacks


CHANGELOG.md
## 1.3.0

* Support custom key widget

## 1.2.0

* Support set divider

## 1.1.0

* Support set height

## 1.0.0

* Support some defaults keyboard layout as well as custom layout


lib/src/piano_widget.dart
import 'package:flutter/widgets.dart';
import 'package:piano_widget/src/key_widget.dart';

import 'octave_widget.dart';

const octave = [
  'C',
  'C#',
  'D',
  'D#',
  'E',
  'F',
  'F#',
  'G',
  'G#',
  'A',
  'A#',
  'B',
];

extension ListExt<T> on List<T> {
  /// Maps each element and its index to a new value.
  Iterable<R> mapIndexed<R>(R Function(int index, T element) convert) sync* {
    for (var index = 0; index < length; index++) {
      yield convert(index, this[index]);
    }
  }
}

class PianoWidget extends StatelessWidget {
  final double height;
  final double divider;
  final List<Iterable<MapEntry<String, int>>> octaves;
  final int whiteCount;
  final KeyBuilder keyBuilder;

  /// The layout with 32 keys and start with C
  factory PianoWidget.keys32c({
    double height = 56,
    double divider = 0.5,
    keyBuilder = buildKey,
  }) {
    return PianoWidget.custom(
      keys: 32,
      startWith: 'C',
      height: height,
      divider: divider,
      keyBuilder: keyBuilder,
    );
  }

  /// The layout with 32 keys and start with F
  factory PianoWidget.keys32f({
    double height = 56,
    double divider = 0.5,
    keyBuilder = buildKey,
  }) {
    return PianoWidget.custom(
      keys: 32,
      startWith: 'F',
      height: height,
      divider: divider,
      keyBuilder: keyBuilder,
    );
  }

  /// The layout with 36 keys and start with C
  factory PianoWidget.keys36c({
    double height = 56,
    double divider = 0.5,
    keyBuilder = buildKey,
  }) {
    return PianoWidget.custom(
      keys: 36,
      startWith: 'C',
      height: height,
      divider: divider,
      keyBuilder: keyBuilder,
    );
  }

  /// The layout with 36 keys and start with F
  factory PianoWidget.keys36f({
    double height = 56,
    double divider = 0.5,
    keyBuilder = buildKey,
  }) {
    return PianoWidget.custom(
      keys: 36,
      startWith: 'F',
      height: height,
      divider: divider,
      keyBuilder: keyBuilder,
    );
  }

  /// The layout with 37 keys and start with C
  factory PianoWidget.keys37c({
    double height = 56,
    double divider = 0.5,
    keyBuilder = buildKey,
  }) {
    return PianoWidget.custom(
      keys: 37,
      startWith: 'C',
      height: height,
      divider: divider,
      keyBuilder: keyBuilder,
    );
  }

  /// The layout with 37 keys and start with F
  factory PianoWidget.keys37f({
    double height = 56,
    double divider = 0.5,
    keyBuilder = buildKey,
  }) {
    return PianoWidget.custom(
      keys: 37,
      startWith: 'F',
      height: height,
      divider: divider,
      keyBuilder: keyBuilder,
    );
  }

  /// The layout with 49 keys
  factory PianoWidget.keys49({
    double height = 56,
    double divider = 0.5,
    keyBuilder = buildKey,
  }) {
    return PianoWidget.custom(
      keys: 49,
      startWith: 'C',
      height: height,
      divider: divider,
      keyBuilder: keyBuilder,
    );
  }

  /// The layout with 54 keys
  factory PianoWidget.keys54({
    double height = 56,
    double divider = 0.5,
    keyBuilder = buildKey,
  }) {
    return PianoWidget.custom(
      keys: 54,
      startWith: 'C',
      height: height,
      divider: divider,
      keyBuilder: keyBuilder,
    );
  }

  /// The layout with 61 keys
  factory PianoWidget.keys61({
    double height = 56,
    double divider = 0.5,
    keyBuilder = buildKey,
  }) {
    return PianoWidget.custom(
      keys: 61,
      startWith: 'C',
      height: height,
      divider: divider,
      keyBuilder: keyBuilder,
    );
  }

  /// The layout with 76 keys
  factory PianoWidget.keys76({
    double height = 56,
    double divider = 0.5,
    keyBuilder = buildKey,
  }) {
    return PianoWidget.custom(
      keys: 76,
      startWith: 'E',
      height: height,
      divider: divider,
      keyBuilder: keyBuilder,
    );
  }

  /// The layout with 88 keys
  factory PianoWidget.keys88({
    double height = 56,
    double divider = 0.5,
    keyBuilder = buildKey,
  }) {
    return PianoWidget.custom(
      keys: 88,
      startWith: 'A',
      height: height,
      divider: divider,
      keyBuilder: keyBuilder,
    );
  }

  factory PianoWidget.custom({
    required int keys,
    required String startWith,
    double height = 56,
    divider = 0.5,
    keyBuilder = buildKey,
  }) {
    final startIndex = octave.indexOf(startWith);
    final firstOctave = startIndex == 0 ? 0 : 12 - startIndex;
    final fullOctaves = (keys - firstOctave) ~/ 12;
    final lastOctave = keys - firstOctave - fullOctaves * 12;
    var whiteCount = fullOctaves * 7;
    if (firstOctave > 0) {
      whiteCount += octave
          .sublist(startIndex, 12)
          .where((element) => !element.contains('#'))
          .length;
    }
    if (lastOctave > 0) {
      whiteCount += octave
          .sublist(0, lastOctave)
          .where((element) => !element.contains('#'))
          .length;
    }
    return PianoWidget._(
      octaves: [
        if (startIndex > 0) // First octave
          octave
              .sublist(startIndex, 12)
              .mapIndexed((index, element) => MapEntry(element, index)),
        for (var i = 0; i < fullOctaves; i++)
          octave.mapIndexed((index, element) {
            return MapEntry(element, firstOctave + i * 12 + index);
          }),
        if (lastOctave > 0)
          octave.sublist(0, lastOctave).mapIndexed(
              (index, element) => MapEntry(element, keys - 1 - index)),
      ],
      whiteCount: whiteCount,
      height: height,
      divider: divider,
      keyBuilder: keyBuilder,
    );
  }

  const PianoWidget._({
    Key? key,
    required this.octaves,
    required this.whiteCount,
    required this.height,
    required this.divider,
    this.keyBuilder = buildKey,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      height: height,
      child: LayoutBuilder(
        builder: (BuildContext context, BoxConstraints constraints) {
          final width = constraints.maxWidth;
          final whiteNoteWidth =
              (width - divider * (whiteCount - 1)) / whiteCount;
          return Row(
            children: [
              for (final widgets in octaves.mapIndexed((index, element) => [
                    OctaveWidget(
                      whiteKeyWidth: whiteNoteWidth,
                      height: height,
                      divider: divider,
                      pitches: element.toList(),
                      keyBuilder: keyBuilder,
                    ),
                    if (index < octaves.length - 1) SizedBox(width: divider),
                  ]))
                ...widgets,
            ],
          );
        },
      ),
    );
  }
}


lib/src/octave_widget.dart
import 'package:flutter/widgets.dart';

import 'key_widget.dart';

Widget buildKey(double width, double height, MapEntry<String, int> pitch) {
  return KeyWidget(width: width, height: height, pitch: pitch);
}

class OctaveWidget extends StatelessWidget {
  final double whiteKeyWidth;
  final double height;
  final double divider;
  final List<MapEntry<String, int>> pitches;
  final KeyBuilder keyBuilder;

  const OctaveWidget({
    Key? key,
    required this.whiteKeyWidth,
    required this.height,
    required this.pitches,
    required this.divider,
    this.keyBuilder = buildKey,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final count = pitches.length;
    final whites = pitches.where((e) => !e.key.contains('#')).toList();
    final whiteCount = whites.length;
    final width = whiteCount * whiteKeyWidth + (whiteCount - 1) * divider;
    return SizedBox(
      width: width,
      height: height,
      child: Stack(
        children: [
          for (var i = 0; i < count; i++) ...[
            if (!pitches[i].key.contains('#'))
              Builder(builder: (context) {
                final index = whites.indexOf(pitches[i]);
                return Positioned.fromRect(
                  rect: Rect.fromLTWH(
                    index * (whiteKeyWidth + divider),
                    0,
                    whiteKeyWidth,
                    height,
                  ),
                  child: keyBuilder(whiteKeyWidth, height, whites[index]),
                );
              }),
            if (!pitches[i].key.contains('#') &&
                i > 0 &&
                pitches[i - 1].key.contains('#'))
              Builder(builder: (context) {
                final index = whites.indexOf(pitches[i]);
                return Positioned.fromRect(
                  rect: Rect.fromCenter(
                    center:
                        Offset(index * (whiteKeyWidth + divider), height / 3),
                    width: whiteKeyWidth * 2 / 3,
                    height: height * 2 / 3,
                  ),
                  child: keyBuilder(whiteKeyWidth, height, pitches[i - 1]),
                );
              })
          ]
        ],
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:piano_widget/piano_widget.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: const Text('Piano'),
      ),
      backgroundColor: Colors.grey,
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: ListView(
          padding: const EdgeInsets.all(16),
          children: [
            const Text('32 Keys start with C, custom clickable key widget'),
            PianoWidget.keys32c(keyBuilder: buildKey),
            const Divider(),
            const Text('32 Keys start with F, wrapper by stateful widget'),
            const MyStatefulWidget(),
            const Divider(),
            const Text('36 Keys start with C'),
            PianoWidget.keys36c(),
            const Divider(),
            const Text('36 Keys start with F'),
            PianoWidget.keys36f(),
            const Divider(),
            const Text('37 Keys start with C'),
            PianoWidget.keys37c(),
            const Divider(),
            const Text('37 Keys start with F'),
            PianoWidget.keys37f(),
            const Divider(),
            const Text('49 Keys'),
            PianoWidget.keys49(),
            const Divider(),
            const Text('54 Keys'),
            PianoWidget.keys54(),
            const Divider(),
            const Text('61 Keys'),
            PianoWidget.keys61(),
            const Divider(),
            const Text('76 Keys'),
            PianoWidget.keys76(),
            const Divider(),
            const Text('88 Keys'),
            PianoWidget.keys88(),
          ],
        ),
      ),
    );
  }

  Widget buildKey(double width, double height, MapEntry<String, int> pitch) {
    return MyKeyWidget(width: width, height: height, pitch: pitch);
  }
}

class MyKeyWidget extends StatefulWidget {
  final double width;
  final double height;
  final MapEntry<String, int> pitch;

  const MyKeyWidget({
    required this.width,
    required this.height,
    required this.pitch,
    Key? key,
  }) : super(key: key);

  @override
  State<MyKeyWidget> createState() => _MyKeyWidgetState();
}

class _MyKeyWidgetState extends State<MyKeyWidget> {
  bool isPressed = false;

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTapDown: (_) {
        setState(() {
          isPressed = true;
        });
      },
      onTapUp: (_) {
        setState(() {
          isPressed = false;
        });
      },
      onTapCancel: () {
        setState(() {
          isPressed = false;
        });
      },
      child: Container(
        width: widget.width,
        height: widget.height,
        decoration: BoxDecoration(
          color: isPressed
              ? Colors.red
              : (widget.pitch.key.contains('#') ? Colors.black : Colors.white),
          borderRadius: const BorderRadius.only(
            bottomLeft: Radius.circular(2),
            bottomRight: Radius.circular(2),
          ),
        ),
      ),
    );
  }
}

class MyStatefulWidget extends StatefulWidget {
  const MyStatefulWidget({Key? key}) : super(key: key);

  @override
  State<MyStatefulWidget> createState() => _MyStatefulWidgetState();
}

class _MyStatefulWidgetState extends State<MyStatefulWidget> {
  var _pressedNotes = <String>{};

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        PianoWidget.keys32f(
          keyBuilder: (width, height, pitch) {
            return Container(
              width: width,
              height: height,
              decoration: BoxDecoration(
                color: _pressedNotes
                        .contains('${pitch.key}${pitch.value ~/ 12 + 1}')
                    ? Colors.red
                    : (pitch.key.contains('#') ? Colors.black : Colors.white),
                borderRadius: const BorderRadius.only(
                  bottomLeft: Radius.circular(2),
                  bottomRight: Radius.circular(2),
                ),
              ),
            );
          },
        ),
        Row(
          children: [
            _buildButton('B2'),
            _buildButton('B3'),
            _buildButton('F#1'),
          ],
        )
      ],
    );
  }

  Widget _buildButton(String note) {
    return GestureDetector(
      onTapDown: (_) {
        _pressedNotes.add(note);
        setState(() {
          _pressedNotes = Set.of(_pressedNotes);
        });
      },
      onTapUp: (_) {
        _pressedNotes.remove(note);
        setState(() {
          _pressedNotes = Set.of(_pressedNotes);
        });
      },
      onTapCancel: () {
        _pressedNotes.remove(note);
        setState(() {
          _pressedNotes = Set.of(_pressedNotes);
        });
      },
      child: Padding(
        padding: const EdgeInsets.all(16.0),
        child: Text(note),
      ),
    );
  }
}
